lib/go-atscfg/remapdotconfig.go

```diff
@@ -0,0 +1,369 @@
+	EdgeHeaderRewrite        *string
+	SigningAlgorithm         *string
+	Name                     string
+	QStringIgnore            *int
+	RegexRemap               *string
+	FQPacingRate             *int
+	DSCP                     int
+	RoutingName              *string
+	MultiSiteOrigin          *string
+	Pattern                  *string
+	RegexType                *string
+	Domain                   *string
+	RegexSetNumber           *string
+	OriginShield             *string
+	ProfileID                *int
+	Protocol                 *int
+	AnonymousBlockingEnabled *bool
+	Active                   bool
+}
+
+func MakeRemapDotConfig(
+	serverName tc.CacheName,
+	toToolName string, // tm.toolname global parameter (TODO: cache itself?)
+	toURL string, // tm.url global parameter (TODO: cache itself?)
+	atsMajorVersion int,
+	cacheURLConfigParams map[string]string, // map[name]value for this server's profile and config file 'cacheurl.config'
+	dsProfilesCacheKeyConfigParams map[int]map[string]string, // map[profileID][paramName]paramVal for this server's profile, config file 'cachekey.config'
+	serverPackageParamData map[string]string, // map[paramName]paramVal for this server, config file 'package'
+	serverInfo *ServerInfo, // ServerInfo for this server
+	remapDSData []RemapConfigDSData,
+) string {
+	hdr := GenericHeaderComment(string(serverName), toToolName, toURL)
+	text := ""
+	if tc.CacheTypeFromString(serverInfo.Type) == tc.CacheTypeMid {
+		text = GetServerConfigRemapDotConfigForMid(atsMajorVersion, dsProfilesCacheKeyConfigParams, serverInfo, remapDSData, hdr)
+	} else {
+		text = GetServerConfigRemapDotConfigForEdge(cacheURLConfigParams, dsProfilesCacheKeyConfigParams, serverPackageParamData, serverInfo, remapDSData, atsMajorVersion, hdr)
+	}
+	return text
+}
+
+func GetServerConfigRemapDotConfigForMid(
+	atsMajorVersion int,
+	profilesCacheKeyConfigParams map[int]map[string]string,
+	server *ServerInfo,
+	dses []RemapConfigDSData,
+	header string,
+) string {
+	midRemaps := map[string]string{}
+	for _, ds := range dses {
+		if ds.Type.IsLive() && !ds.Type.IsNational() {
+			continue // Live local delivery services skip mids
+		}
+
+		if ds.OriginFQDN == nil || *ds.OriginFQDN == "" {
+			log.Warnf("GetServerConfigRemapDotConfigForMid ds '" + ds.Name + "' has no origin fqdn, skipping!") // TODO confirm - Perl uses without checking!
+			continue
+		}
+
+		if midRemaps[*ds.OriginFQDN] != "" {
+			continue // skip remap rules from extra HOST_REGEXP entries
+		}
+
+		midRemap := ""
+		if ds.MidHeaderRewrite != nil && *ds.MidHeaderRewrite != "" {
+			midRemap += ` @plugin=header_rewrite.so @pparam=` + MidHeaderRewriteConfigFileName(ds.Name)
+		}
+		if ds.QStringIgnore != nil && *ds.QStringIgnore == tc.QueryStringIgnoreIgnoreInCacheKeyAndPassUp {
+			midRemap += GetQStringIgnoreRemap(atsMajorVersion)
+		}
+		if ds.CacheURL != nil && *ds.CacheURL != "" {
+			midRemap += ` @plugin=cacheurl.so @pparam=` + CacheURLConfigFileName(ds.Name)
+		}
```

Comment:
Right now, multiple uses of `cacheurl` do not do what one might expect. It's probably worth actually enhancing this to make these duplications findable. We should probably fix `cacheurl` as well, but it's basically never right to have two of these on a line at the same time.

It looks like we handle it for the edge case, but it is just as problematic at the mid.

Reply:
Ok, I added error logging any time the config gen adds multiple cachekey or cacheurl entries on the same line.